node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/NodeAlerter.java

```diff
@@ -0,0 +1,442 @@
+        return hosts.stream()
+                .filter(h -> h.state() != Node.State.failed && h.state() != Node.State.parked)
+                .collect(Collectors.toList());
+    }
+
+    private Optional<HostFailurePath> greedyHeuristicFindFailurePath(Map<Node, Integer> heuristic, List<Node> hosts,
+                                                   Map<Node, List<Node>> nodeChildren,
+                                                   Map<Node, AllocationResources> availableResources) {
+        if (hosts.size() == 0) return Optional.empty();
+        List<Node> parentRemovalPriorityList = heuristic.entrySet().stream()
+                .sorted(Comparator.comparingInt(Map.Entry::getValue))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+        for (int i = 1; i <= parentRemovalPriorityList.size(); i++) {
+            List<Node> hostsToRemove = parentRemovalPriorityList.subList(0, i);
+            var hostRemovalFailure = findHostRemovalFailure(hostsToRemove, hosts, nodeChildren, availableResources);
+            if (hostRemovalFailure.isPresent()) {
+                HostFailurePath failurePath = new HostFailurePath();
+                failurePath.hostsCausingFailure = hostsToRemove;
+                failurePath.failureReason = hostRemovalFailure.get();
+                return Optional.of(failurePath);
+            }
+        }
+
+        throw new IllegalStateException("No path to failure found. This should be impossible!");
+    }
+
+    int countOvercommittedHosts() {
+        List<Node> hosts = filterHosts(nodeRepository.getNodes(NodeType.host));
+        List<Node> tenants = filterTenants(nodeRepository.getNodes(NodeType.tenant), hosts);
+        var nodeMap = constructHostnameToNodeMap(hosts);
+        var nodeChildren = constructNodeChildrenMap(tenants, hosts, nodeMap);
+        var availableResources = constructAvailableResourcesMap(hosts, nodeChildren);
+
+        List<Node> overcommittedNodes = findOvercommittedNodes(availableResources);
+        if (overcommittedNodes.size() != 0) {
+            log.log(LogLevel.WARNING, String.format("%d nodes are overcommitted! [ %s ]", overcommittedNodes.size(),
+                    overcommittedNodes.stream().map(Node::hostname).collect(Collectors.joining(", "))));
+        }
+        return overcommittedNodes.size();
+    }
+
+    private Map<String, Node> constructHostnameToNodeMap(List<Node> nodes) {
+        return nodes.stream().collect(Collectors.toMap(Node::hostname, n -> n));
+    }
+
+    private Map<Node, List<Node>> constructNodeChildrenMap(List<Node> tenants, List<Node> hosts, Map<String, Node> hostnameToNode) {
+        Map<Optional<Node>, List<Node>> possibleNodeChildren = tenants.stream()
+                .filter(n -> n.parentHostname().isPresent())
+                .collect(Collectors.groupingBy(
+                        n -> Optional.ofNullable(hostnameToNode.get(n.parentHostname().orElseThrow()))));
```

Comment:
If the `hostnameToNode` map does not contain an entry for the listed parent host name, the result will be `null`. GroupingBy throws a `NullPointerException` if attempting to group elements under the key `null`, even though this is valid behaviour for a normal hashmap.

I used the size of the removed children found in the next line at some point, but since that's no longer the case, I could update the code to the following;
```Java
        Map<Node, List<Node>> nodeChildren = tenants.stream()
                .filter(n -> n.parentHostname().isPresent())
                .filter(n -> hostnameToNode.containsKey(n.parentHostname().get()))
                .collect(Collectors.groupingBy(
                        n -> hostnameToNode.get(n.parentHostname().orElseThrow())));

        for (var host : hosts) nodeChildren.putIfAbsent(host, List.of());

        return nodeChildren;
```

Reply:
If the parent hostname is not in the `hostnameToNode` map, we are in a completely borked state, and we shouldn't continue.